### DISCUSSION TITLE
jxMapviewer2 Discussions!

### DISCUSSION THREAD

**Discussion by msteiger** [msteiger/jxmapviewer2#134] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

**Comment by garciarodrigo99** [comment]
Good evening. I want to work with offline maps, and I would like to know how must be the format to work, due to in [https://github.com/msteiger/jxmapviewer2/blob/master/examples/src/sample9_offline/Sample9.java ](url) indicates it must be a zip following a specific path. Two questions:

What would the structure of the format be like?
Is it possible to work with .osm.pbf format?
INFO: Removing from map: file:///home/hide_path/src/main/resources/maps/spain_merge.osm.pbf&null=2049&null=2046&null=5, tile failed to load
abr 01, 2024 1:36:42 P. M. org.jxmapviewer.viewer.AbstractTileFactory run
SEVERE: Unable to load tile: /home/hide_path/src/main/resources/maps/spain_merge.osm.pbf&null=2048&null=2049&null=5 ((The file or directory does not exist)
abr 01, 2024 1:36:42 P. M. org.jxmapviewer.viewer.AbstractTileFactory run
SEVERE: Unable to load tile: /home/hide_path/src/main/resources/maps/spain_merge.osm.pbf&null=1024&null=1024&null=6 ((The file or directory does not exist)
abr 01, 2024 1:36:42 P. M. org.jxmapviewer.viewer.AbstractTileFactory getTile
The file exists and the route is correct, before that the content of the file is write on console.
Here the code:
`
public class MapViewerApp
{
    public static void main(String[] args) throws IOException
    {
        // Ruta al archivo .osm.pbf
        String osmFileLocalPath = "src/main/resources/maps/spain_merge.osm.pbf";
        File osmFile = new File(osmFileLocalPath);

        // Construir la información de la fábrica de azulejos
        TileFactoryInfo info = new TileFactoryInfo(0, // min zoom level
                17, // max zoom level
                17, // total number of zoom levels
                256, // tile size
                true, // x/y orientation is normal
                true, // tile images are square
                "file://"+osmFile.getAbsolutePath(), // base URL
                null, null, null);

        // Crear la fábrica de azulejos
        TileFactory tileFactory = new DefaultTileFactory(info);

        // Setup JXMapViewer
        final JXMapViewer mapViewer = new JXMapViewer();
        mapViewer.setTileFactory(tileFactory);

        // Centrar el mapa en una posición predeterminada
        GeoPosition center = new GeoPosition(0, 0); // Puedes ajustar estas coordenadas según sea necesario
        mapViewer.setZoom(5); // Puedes ajustar el nivel de zoom inicial según sea necesario
        mapViewer.setAddressLocation(center);

        // Add interactions
        MouseInputListener mia = new PanMouseInputListener(mapViewer);
        mapViewer.addMouseListener(mia);
        mapViewer.addMouseMotionListener(mia);
        mapViewer.addMouseWheelListener(new ZoomMouseWheelListenerCursor(mapViewer));

        // Display the viewer in a JFrame
        final JFrame frame = new JFrame();
        frame.setLayout(new BorderLayout());
        frame.add(new JLabel("Use left mouse button to pan and mouse wheel to zoom"), BorderLayout.NORTH);
        frame.add(mapViewer);
        frame.setSize(800, 600);
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setTitle("JXMapviewer2 Example 9 - Offline maps");
        frame.setVisible(true);
    }
}
`
Thank you so much.

**Reply by msteiger** [reply]
Currently, the structure inside the zip file has to be exactly the same that is used for caching. This makes it very easy to create offline maps. Just zip the cache folder.

That said, I believe that it would be awesome to be able to process OSM files in pbf format. Do you think that you contribute to that? I would assume that there's already a Java library to parse the files, so it shouldn't be super-hard.

**Comment by heavydriver** [comment]
How do use MapTiler tiles with JXMapViewer2.

This is the url:
https://api.maptiler.com/maps/streets-v2/{z}/{x}/{y}.png?key={API_KEY}
